lib/lib.es5.d.ts

```diff
@@ -1408,7 +1408,7 @@ interface ArrayConstructor {
     (arrayLength?: number): any[];
     <T>(arrayLength: number): T[];
     <T>(...items: T[]): T[];
-    isArray(arg: any): arg is any[];
+    isArray(arg: any): arg is readonly unknown[];
```

Comment:
This is a safer type, but it can't help but break anybody who was previously able to mutate an array or use its contents after checking it with isArray. I think it's too breaky for the benefit.

Reply:
@sandersn You mean anybody who was previously able to mutate a mutable array after checking it with isArray? I tested this change locally and confirmed it doesn't break that: Mutable arrays remain mutable. The type predicate narrows the type to the intersection of `typeof arg & readonly unknown[]`, which is only immutable if `arg` was already immutable.

Would you consider reopening this PR?

### :play_or_pause_button: Playground link

<!--
  A link to a TypeScript Playground "Share" link which shows this behavior

  The TypeScript Workbench can be used for more complex setups, try
  https://www.typescriptlang.org/dev/bug-workbench/

  As a last resort, you can link to a repo, but these will be slower for us to investigate.
-->
[Workbench repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwK4agCMIQAueTGASxQHMBtAXQG4AoUSWBZdLK7PAWJl4cKMFQQAnhQzU6TNqwD0y+AFl8REvFgwoUtKJDYoNCkngB3KhgAWulMGu27SfCrV2MGAA5pSVVpXXEIAOmRsZWwqMBgkNCQAMwxlABUpXxAAZTiqX1TfXAgIZQAWACYAZgBGADZWKiT4AAoAQRh9KTCqNA6ulsFtEABKEfgAb1Z4RFRMCjdi4HJKGgZGeABeHC1hZnhVeEAUckB4P-g2naEdPQMF9whnOFNzRNYAX1ZPeDS7Xt00NBUWgobAgFBYNCLB7wECdJAweCEfDwCQgNAoADkGAANPAUEgsDZ7O4MF9vH4AkEQuFItFYvFEil0pkcnkCsoiiVytV6o1mu1OgYen1BVIWvwhsIxpNprNeHclgA5AkrORrJhbeAS3YkfaHQAy5OcUFqBDqEDcZJD7o8TGZja83kA).

### :technologist: Code

<!-- Please post the relevant code sample here as well -->
```ts repro
declare const mutable: string[];
declare const immutable: readonly string[];

// Mutable arrays remain so with and without
// https://github.com/microsoft/TypeScript/pull/42316
if (Array.isArray(mutable)) {
  const should: string[] = mutable; // ✔️ A mutable array should remain so
}

// This assignment should error but doesn't, not without
// https://github.com/microsoft/TypeScript/pull/42316
if (Array.isArray(immutable)) {
  const shouldNot: string[] = immutable; // ❌ An immutable array should remain so
}
```

### :slightly_frowning_face: Actual behavior

<!-- What happened, and why it was wrong -->
```sh
$ tsc input.ts
```
No errors: You're free to mutate an immutable array.

### :slightly_smiling_face: Expected behavior

<!-- What you expected to happen instead, and why -->
With this PR:
```sh
$ node built/local/tsc.js input.ts 
input.ts:13:9 - error TS4104: The type 'readonly string[]' is 'readonly' and cannot be assigned to the mutable type 'string[]'.

13   const shouldNot: string[] = immutable; // ❌ An immutable array should remain so
           ~~~~~~~~~


Found 1 error in input.ts:13
```
Mutating an immutable array is an error.

Reply:
What I mean is that if somebody writes this today, it's not an error:

```ts
declare const u: unknown;
if (Array.isArray(u)) {
    u[1] = 12
    u[1].length
}
```

`unknown` narrowing is an important scenario for Array.isArray, and this change would add errors that were not there before and are not necessarily correct.

Reply:
@sandersn I should've realized what you meant, thanks for your patience.

Instead of replacing the existing signature, would adding overloads for `ArrayLike<T>` and `Iterable<T>` work, without affecting the current behavior of `any` and `unknown`? I've opened https://github.com/microsoft/TypeScript/pull/48228 with that modification to this PR.